src/extra/components/xgm.py

```diff
@@ -0,0 +1,353 @@
+                with XGMs on SASE 1 or SASE 3 and you're only interested in data
+                for a specific SASE. This setting applies to all the
+                SASE-specific fields of an XGM, such as
+                [XGM.intensity()][extra.components.XGM.intensity] and
+                [XGM.npulses()][extra.components.XGM.npulses]. If it is not set
+                the class will read the main properties of the XGM,
+                e.g. `data.intensityTD` instead of `data.intensitySa1TD`.
+        """
+        # If the user doesn't choose a device, attempt to find one
+        if device is None:
+            xgms = [source for source in run.control_sources
+                    if run[source].device_class in self.DEVICE_CLASSES]
+            if len(xgms) == 0:
+                xgms = [source for source in run.control_sources
+                        if "XGM/DOOCS" in source or "XGMD/DOOCS" in source]
+
+            if len(xgms) == 0:
+                raise RuntimeError("Couldn't find an XGM in the run, please pass the device name with the `device` argument")
+            elif len(xgms) > 1:
+                raise RuntimeError(f"Multiple XGMs found, please select one with the `device` argument: {xgms}")
+            elif len(xgms) == 1:
+                device = xgms[0]
+        elif isinstance(device, KeyData):
+            # If a KeyData is passed, assume it belongs to an XGM source
+            device = device.source
+
+        # We put this last part of argument checking in a separate if-block so that if a KeyData was passed
+        # and it assigns a SourceData object, we enter the SourceData branch below.
+        if isinstance(device, SourceData):
+            # If it's an instrument source, assume it's the output channel
+            if device.is_instrument and device.source.endswith(":output"):
+                device = device.source.removesuffix(":output")
+            else:
+                device = device.source
+        elif isinstance(device, str):
+            pass
+        else:
+            raise TypeError(f"Unrecognized `device` type '{type(device)}', must be a SourceData, KeyData, or a string")
+
+        self._default_pg = PropertyGroup(default_sase) if default_sase is not None else PropertyGroup.MAIN
+
+        self._control_source = run[device]
+        self._instrument_source = run[f"{device}:output"]
+
+        self._intensity = { }
+        self._npulses = { }
+        self._mean_pulses = { }
+        self._max_pulses = { }
+        self._photon_flux = None
+
+        self._extra_sases = [sase for sase in [1, 3]
+                             if f"pulseEnergy.numberOfSa{sase}BunchesActual" in self._control_source]
+
+        from scipy.constants import h, c, e
+        self._wavelength = self._control_source["pulseEnergy.wavelengthUsed"].as_single_value()
+        self._photon_energy = (h * c / e) / (self._wavelength * 1e-6)
+
+        self._doocs_server = self._control_source.run_values()["location.value"]
```

Comment:
I would also move this out of the init, as this would fail if we try to instantiate the object on an Unioned DataCollection, and it looks like it's not used anywhere else that the `doocs_server` property.

Now that I think about it, it may be worth adding a test with combined runs for this component :p